File: 257-BinaryTreePaths/257-BinaryTreePaths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    ArrayList<String> ans = new ArrayList();
    public List<String> binaryTreePaths(TreeNode root) {
        if (root != null) {
            dfs(root, new StringBuilder(""));
        }
        return ans;
    }

    public void dfs(TreeNode root, StringBuilder sb) {
        if (root == null) {
            return;
        }

        int length = sb.length();

        if (length > 0) {
            sb.append("->");
        }
        sb.append(root.val);

        if (root.left == null && root.right == null) {
            ans.add(sb.toString());
        } else {
            dfs(root.left, sb);
            dfs(root.right, sb);
        }

        sb.setLength(length);
    }
}

File: 257-BinaryTreePaths/257-BinaryTreePaths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[str]
        """
        ans = []

        def helper(root, path):
            if not root:
                return
            
            if not root.left and not root.right:
                ans.append(path + str(root.val))
            
            if root.left:
                helper(root.left, path + str(root.val) + "->")
            if root.right:
                helper(root.right, path + str(root.val) + "->")
        
        helper(root, "")
        return ans